r2.py

```diff
@@ -0,0 +1,299 @@
+            result = node(*grads)
+            return result
+        finally:
+            swap_saved_values(node, new_saved_values)
+    return apply_node_with_saved
+
+
+def is_accumulate_grad(node):
+    return node.__class__.__name__ == "AccumulateGrad"
+
+
+def visit_node(node, visit_saved, visit_param=None):
+    for attr in dir(node):
+        if attr.startswith("_raw"):
+            old_saved_value = getattr(node, attr)
+            visit_saved(old_saved_value)
+    if is_accumulate_grad(node):
+        if visit_param is not None:
+            visit_param(node.variable)
+
+
+def swap_saved_values(node, saved):
+    counter = 0
+
+    def visit(old_saved_value):
+        nonlocal counter
+        if counter >= len(saved):
+            breakpoint()
+        new_saved_value = saved[counter]
+        torch._C._swap_saved(old_saved_value, new_saved_value)
+        counter += 1
+
+    visit_node(node, visit)
+
+
+def num_saved(node):
+    counter = 0
+
+    def visit_saved(val):
+        nonlocal counter
+        counter += 1
+
+    def visit_param(param):
+        nonlocal counter
+        counter += 1
+
+    visit_node(node, visit_saved, visit_param)
+    return counter
+
+
+def construct_backward_function(nodes):
+    # Do some preprocessing here
+    apply_with_saved = {idx: create_apply_node_with_saved(node) for idx, node in enumerate(nodes)}
+    num_saved_values = {idx: num_saved(node) for idx, node in enumerate(nodes)}
+    node_to_idx = {node: idx for idx, node in enumerate(nodes)}
+    next_functions = {
+        idx: tuple((node_to_idx[next_node], next_idx) for next_node, next_idx in node.next_functions)
+        for idx, node in enumerate(nodes)
+    }
+    is_accumulate_grad_node = {idx: is_accumulate_grad(node) for idx, node in enumerate(nodes)}
+    pre_hooks = {idx: node.pre_hooks() for idx, node in enumerate(nodes)}
+
+    def execute_autograd(input_buffer, saved_values):
+        """Captures the autograd nodes to be executed.
+        This function is essentially a re-implementation of the autograd engine
+        in Python.
+        """
+        all_input_buffers = {0: [input_buffer]}
+        saved_values_begin = 0
+        for idx, node in enumerate(nodes):
+            input_buffers = all_input_buffers[idx]
+
+            saved_values_end = saved_values_begin + num_saved_values[idx]
+
+            # We support prehooks.
+            if len(pre_hooks[idx]) > 0:
+                assert len(pre_hooks[idx][0].values()) == 1
+                pre_hook = list(pre_hooks[idx][0].values())[0]
+                input_buffers = pre_hook(input_buffers)
+
+            if is_accumulate_grad_node[idx]:
+                # Rewrite AccumulateGrad nodes into a special op.
+                assert len(input_buffers) == 1
+                param = saved_values[saved_values_begin]
+                param_grad = input_buffers[0].expand_as(param)
+                grad_inputs = [torch.ops.inductor.accumulate_grad_.default(param, param_grad)]
```

Comment:
we have one fx pass in compiled autograd that must happen before dynamo: https://github.com/pytorch/pytorch/blob/ceb2fcc5db5926d6f57adb5954424953ad5e0fc0/torch/_dynamo/compiled_autograd.py#L439-L444

Reply:
This is for the compiled autograd over a torch.compile region case right? I expect that we need special case handling for this situation; somehow the torch.compile region needs to expose the AOTAutograd backward graph to us and then we can run the reorder_accumluate_grad_nodes pass on it.

Reply:
Ah we have a reference to the post grad graph stored in the PyNode